### PR TITLE
fixed to play games

### PR DESCRIPTION
**Pull request by mkorganashvili**
So, here are some fixes to play live and archive games.

The site has changed a lot and I tried to fix it with minimal effort, spending more times on digging details of video streaming. So the code is not as pretty as original and misses some minor features, like sorting video options and so on.

The big problem is, that it only works on windows. Seems like issue of widevine and I am working hard to find a solution and make it work

### CODE REVIEW

**Comment by RHStonks** [comment]
@mkorganashvili cheers. I thought updating this addon won't be possible but it seems like you have done it. 

Can't wait to see you update live games.

**Comment by mkorganashvili** [comment]
> 

Cheers @RHStonks. 
The bad part is that it only works on windows but let's see : )

**Comment by RHStonks** [comment]
> > 
> 
> 
> 
> Cheers @RHStonks. 
> 
> The bad part is that it only works on windows but let's see : )

Great stuff. Glad to see that after a week you also got live content going.

So does that mean it's not working on Android & Android TV when you mentioned Windows only?

**Comment by mkorganashvili** [comment]
> So does that mean it's not working on Android & Android TV when you mentioned Windows only?

Well, I tried Linux and iOS, both failed. I am not sure about Android but don't think it is going to work

**Comment by HoRsEman3GB** [comment]
Had some luck with Android TV on a quick play last night on both Android TV 9 and Android TV 12.

Some archived games worked but not all oddly.

Strangely no luck on windows

Great progress though :)

Appreciate the hard work

**Comment by RHStonks** [comment]
> Had some luck with Android TV on a quick play last night on both Android TV 9 and Android TV 12.
> 
> 
> 
> Some archived games worked but not all oddly.
> 
> 
> 
> Strangely no luck on windows
> 
> 
> 
> Great progress though :)
> 
> 
> 
> Appreciate the hard work

What about live games on Android TV. Did it work for you?

**Comment by chamchenko** [comment]
Thanks a lot for the helping hand, here for linux I doubt it will be possible unless the input stream adaptive team will have some miracle bypass solution to the nba intentionally making their stream not work on linux. the issue has been confirmed from some reddit users [here.](https://www.reddit.com/r/Piracy/comments/xujtdc/thanks_to_microsoft_nba_league_pass_no_longer/) I myself haven't been able to play game on linux, and that's mostly what kept from working on this as I needed to set testing environment on windows or android which I didn't have the time or the nerve to do lately.

**Comment by RHStonks** [comment]
@chamchenko can you push an update to the addon with the current changes that @mkorganashvili so I can test it on both Android and Android TV to see if live games are working?

**Comment by HoRsEman3GB** [comment]
> > Had some luck with Android TV on a quick play last night on both Android TV 9 and Android TV 12.
> > Some archived games worked but not all oddly.
> > Strangely no luck on windows
> > Great progress though :)
> > Appreciate the hard work
> 
> What about live games on Android TV. Did it work for you?

No luck on either of the 2 early games. But I was on windows at work though (I'm in New Zealand) will have look at the Lakers game when get home in an hour and a half and report back

**Comment by chamchenko** [comment]
> @chamchenko can you push an update to the addon with the current changes that @mkorganashvili so I can test it on both Android and Android TV to see if live games are working?

it's done you can now update from the repository

**Comment by RHStonks** [comment]
> > @chamchenko can you push an update to the addon with the current changes that @mkorganashvili so I can test it on both Android and Android TV to see if live games are working?
> 
> 
> 
> it's done you can now update from the repository

@chamchenko I've having issues with installing the repo it says: Failed to install addon from zip file due to invalid structure.

EDIT: Seems installing addon zip file works but installing repo zip file doesn't work. 

I get this problem when opening the addon: Type Error - Not all arguments converted during string formatting on Android.

**Comment by HoRsEman3GB** [comment]
> > > @chamchenko can you push an update to the addon with the current changes that @mkorganashvili so I can test it on both Android and Android TV to see if live games are working?
> > 
> > 
> > it's done you can now update from the repository
> 
> @chamchenko I've having issues with installing the repo it says: Failed to install addon from zip file due to invalid structure.
> 
> EDIT: Seems installing addon zip file works but installing repo zip file doesn't work.
> 
> I get this problem when opening the addon: Type Error - Not all arguments converted during string formatting on Android.

On Android TV 10 dongle right now the DEV version is playing Lakers vs Wizards live for me as we speak.

Patch1 and repo version (which appears to be patch1) do not

The playback is slightly sluggish but in sync at highest res.

Oddly still cannot get to work on windows and I did recently update my widevine on that

**Comment by RHStonks** [comment]
@HoRsEman3GB I'm 100% certain that I downloaded the dev version on my Android tablet. 

How do you download it. I'm on the NBA kodi addon page. Do you press code and download zip? 

Surely there should be no difference between Android & Android TV so it'a strange that it's working for you but not me.

**Comment by HoRsEman3GB** [comment]
Yes that's all I've done.
Can vouch for ex different Android TV devices now. But oddly my phone (Google Pixel 6) doesn't seem to work with any add-ons that require widevine so unable to test on that. I do have a tablet I can dig out and try later.

But the dongles appear to be all go for Live games. I'm just seeing what else works

**Comment by RHStonks** [comment]
> Yes that's all I've done.
> 
> Can vouch for ex different Android TV devices now. But oddly my phone (Google Pixel 6) doesn't seem to work with any add-ons that require widevine so unable to test on that. I do have a tablet I can dig out and try later.
> 
> 
> 
> But the dongles appear to be all go for Live games. I'm just seeing what else works

I do have Android TV device on AndroidTV 9 (Mi Box S) so I will give that a go later since you had success with it. Maybe android phones and tablets don't work.

**Comment by mkorganashvili** [comment]
> the issue has been confirmed from some reddit users [here](https://www.reddit.com/r/Piracy/comments/xujtdc/thanks_to_microsoft_nba_league_pass_no_longer/)

well, this is strange, because I tested on Mac and it plays in chrome but does not work in kodi

**Comment by mkorganashvili** [comment]
> Strangely no luck on windows

@HoRsEman3GB can you share logs?

File: resources/lib/games.py [hunk 1]
```diff
@@ -458,122 +462,63 @@ def BROWSE_YEARS(plugin, year, team=False):
...
 
+@Resolver.register
+def PLAY_GAME(plugin, gameID, videoProfileId):
+    access_token = get_token()
+    deviceinfos = get_device_ids()
+    headers = {
+        'content-type': 'text/plain;charset=UTF-8',
+        'authorizationtoken': access_token,
+        'azukiimc': 'IMC7.1.0_AN_D3.0.0_S0',
+        'deviceprofile': b64encode(('{"osVersion":"10","vendorName":"Microsoft","osName":"HTML5","deviceUUID":"%s"}' % deviceinfos['DEVICEID']).encode('ascii'))
```
Comment (nre-ableton) [inline resources/lib/games.py:491]: This line seems pretty Windows-specific. Is it here to "spoof" a browser/OS combination to the API? Or should we rather report the actual OS version?

Comment (mkorganashvili) [reply resources/lib/games.py:491]: I tried to spoof the API, tried different `deviceprofiles` but did not work

Comment (nre-ableton) [reply resources/lib/games.py:491]: @mkorganashvili Interesting. So you're saying that it _only_ works with the above `deviceprofile`?

Comment (mkorganashvili) [reply resources/lib/games.py:491]: I mean it works with any deviceprofile

**Comment by RHStonks** [comment]
Just to let you know that on Android TV I could get archive games going. Quite sluggish like @HoRsEman3GB said, subtitles not showing correctly and not all games work. 

Tried to watch Rapters vs 76ers or Jazz vs Cavs live and it doesn't work and pops up a Runtime Error: No Items Found

**Comment by HoRsEman3GB** [comment]
> > Strangely no luck on windows
> 
> @HoRsEman3GB can you share logs?

Sorry not in a position to create a log from my work machine (strictly Kodi is a no no in general ;) ) What i can tell you is it says KeyError 'cdns' when selecting any of the the many available game feeds inside of every available game. Live or archived.

The games are working great on Android TV though as discussed which is great at home :)

**Comment by RHStonks** [comment]
@HoRsEman3GB I haven't tried watching live games on Android TV since it wasn't working the other day. Could only get archive games going.

Have you had success with all live games so far on Android TV Kodi?

**Comment by HoRsEman3GB** [comment]
> @HoRsEman3GB I haven't tried watching live games on Android TV since it wasn't working the other day. Could only get archive games going.
> 
> Have you had success with all live games so far on Android TV Kodi?

Yeah I have

Watching my Sixers vs Detroit on Kodi 19.4 on a ChromeCast w Google TV (Android 12) as we speak

**Comment by nre-ableton** [comment]
I tried last night on my Kodi box, which is a raspberry pi 4 running LibreELEC. No luck streaming games, and sometimes I also got `KeyErrors` while browsing archives. Also, all of the archived games were shown as "LIVE". I'll try to debug this a bit over the holidays if I have time.

**Comment by mkorganashvili** [comment]
> I also got `KeyErrors`

All other errors are easily fixable but we are but yet nothing to do with main problem. Decryptor seems to act differently on Kodi.

P.S. I'm using Libreelec as well and no luck :(

**Comment by RHStonks** [comment]
I had archive games going on my Android TV but I thought I may have accidentally installed the no -dev version since @HoRsEman3GB got live games working so I deleted it and reinstall again and now can't watch live games or archive anymore. It pops up with TypeError everytime I open the addon.

**Comment by HoRsEman3GB** [comment]
Just a wee note
Live games no longer working on Kodi 20 for me on Android TV. but Archive are...

**Pull request merged**